bin/opd8n/src/cmd/from_l1.rs

```diff
@@ -114,55 +194,105 @@ impl FromL1 {
+    }
 
-        let blobs = blobs::load(
-            &block_info,
-            txs.as_slice(),
-            batcher_address,
-            signer,
-            blob_provider,
-        )
-        .await?;
+    /// Returns a new [AlloyChainProvider] using the l1 rpc url.
+    pub fn l1_provider(&self) -> Result<AlloyChainProvider> {
+        Ok(AlloyChainProvider::new_http(self.l1_rpc_url()?))
+    }
+
+    /// Returns a new [AlloyL2ChainProvider] using the l2 rpc url.
+    pub fn l2_provider(&self, cfg: Arc<RollupConfig>) -> Result<AlloyL2ChainProvider> {
+        Ok(AlloyL2ChainProvider::new_http(self.l2_rpc_url()?, cfg))
+    }
+
+    /// Returns a new [StatefulAttributesBuilder] using the l1 and l2 providers.
+    pub fn attributes(
+        &self,
+        cfg: Arc<RollupConfig>,
+        l2_provider: &AlloyL2ChainProvider,
+        l1_provider: &AlloyChainProvider,
+    ) -> StatefulAttributesBuilder<AlloyChainProvider, AlloyL2ChainProvider> {
+        StatefulAttributesBuilder::new(cfg, l2_provider.clone(), l1_provider.clone())
+    }
+
+    /// Returns a new [OnlineBlobProviderWithFallback] using the beacon url.
+    pub fn blob_provider(
+        &self,
+    ) -> OnlineBlobProviderWithFallback<OnlineBeaconClient, OnlineBeaconClient, SimpleSlotDerivation>
+    {
+        OnlineBlobProviderBuilder::new()
+            .with_beacon_client(OnlineBeaconClient::new_http(self.beacon_url.clone()))
+            .build()
+    }
+
+    /// Returns a new [EthereumDataSource] using the l1 provider and blob provider.
+    pub fn dap(
+        &self,
+        l1_provider: AlloyChainProvider,
+        blob_provider: OnlineBlobProviderWithFallback<
+            OnlineBeaconClient,
+            OnlineBeaconClient,
+            SimpleSlotDerivation,
+        >,
+        cfg: &RollupConfig,
+    ) -> EthereumDataSource<
+        AlloyChainProvider,
+        OnlineBlobProviderWithFallback<
+            OnlineBeaconClient,
+            OnlineBeaconClient,
+            SimpleSlotDerivation,
+        >,
+    > {
+        EthereumDataSource::new(l1_provider, blob_provider, cfg)
+    }
+
+    /// Gets the rollup config from the l2 rpc url.
+    pub async fn rollup_config(&self) -> Result<RollupConfig> {
+        let mut l2_provider =
+            AlloyL2ChainProvider::new_http(self.l2_rpc_url()?, Arc::new(Default::default()));
+        let l2_chain_id = l2_provider.chain_id().await.map_err(|e| eyre!(e))?;
+        let cfg = ROLLUP_CONFIGS
+            .get(&l2_chain_id)
+            .cloned()
+            .ok_or_else(|| eyre!("No rollup config found for L2 chain ID: {}", l2_chain_id))?;
+        Ok(cfg)
+    }
+
+    /// Returns the l1 rpc url from CLI or environment variable.
+    pub fn l1_rpc_url(&self) -> Result<Url> {
+        Url::parse(&self.l1_rpc_url).map_err(|e| eyre!(e))
+    }
+
+    /// Returns the l2 rpc url from CLI or environment variable.
+    pub fn l2_rpc_url(&self) -> Result<Url> {
+        Url::parse(&self.l2_rpc_url).map_err(|e| eyre!(e))
```

Comment:
Consider simplifying to
```suggestion
  &self.l2_rpc_url.parse()?
```